core/src/main/java/io/fabric8/maven/core/service/openshift/OpenshiftBuildService.java

```diff
@@ -447,21 +417,23 @@ private void waitUntilBuildFinished(CountDownLatch latch) {
         }
     }
 
-    private Watcher<Build> getBuildWatcher(final CountDownLatch latch, final String buildName, final AtomicReference<Build> buildHolder) {
-        return new Watcher<Build>() {
+    private Watcher<Pod> getBuildWatcher(final CountDownLatch latch, final CountDownLatch readyLatch, final String buildName) {
+        return new Watcher<Pod>() {
 
             String lastStatus = "";
 
             @Override
-            public void eventReceived(Action action, Build build) {
-                buildHolder.set(build);
-                String status = KubernetesResourceUtil.getBuildStatusPhase(build);
-                log.verbose("BuildWatch: Received event %s , build status: %s", action, build.getStatus());
+            public void eventReceived(Action action, Pod buildPod) {
```

Comment:
This and the `onClose` method are the points where the latch is released. But does the kubernetes client guarantee that they are called? In the past we added additional because in some scenarios we missed the event when the build was too fast (e.g. the base image does not exist), and the latch was not released.

Plus, the onClose does not mean that the build is finished. It can be fired also when the build is not existent. Previously there was a check on the pod to ensure that we start watching after that the build is started.

Reply:
In my opinion kubernetes-client is ensuring that. In [WatchConnectionManager](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchConnectionManager.java#L225 ) during the OkHttp's websocket `onMessage()` callback, we're calling `watcher.eventReceived` in two cases and  `watcher.onClose` in `onClose()`, `onFailure()` and `onClosing()` callbacks.

You're right about onClose, I'm going to close `readlyLatch` in that block also.